### PR TITLE
refactor(auth)!: State machine facade

### PR DESCRIPTION
**Pull request by dnys1**
- Renames `AuthStateMachine` to `ConfigurationStateMachine`
- Refactors the Cognito state machine facade a bit to decrease coupling between the different sub state machines.
- Adds a distinct queue for the plugin so that public API calls are serially executed whereas internal state machine calls can be concurrently executed to avoid deadlock.
- Consolidates state machine API to encourage proper usage.

### CODE REVIEW

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/credential_store_state_machine.dart [hunk 1]
```diff
@@ -10,44 +10,78 @@ import 'package:amplify_auth_cognito_dart/src/credentials/credential_store_keys.
...
+        return;
+      case CredentialStoreEventType.failed:
+        event as CredentialStoreFailed;
+        emit(CredentialStoreState.failure(event.exception));
+        return;
+    }
+  }
+
+  @override
+  CredentialStoreState? resolveError(Object error, [StackTrace? st]) {
```
Comment (ragingsquirrel3) [inline packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/credential_store_state_machine.dart:80]: same question here regarding `st`

**Review by ragingsquirrel3 — APPROVED** [review]
Approving this knowing that there will be some conflicts from fetchAuthSession PR but seems like resolution won't be too bad. Nice job on refactor, found the newer logic easier to follow and everything seemed to fit together well. I'm sure a lot of thought and revision went into it to make it seem that way.

I left a few minor nits and non-blocking questions/comments.

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/configuration_state_machine.dart [hunk 2]
```diff
@@ -9,32 +9,70 @@ import 'package:amplify_auth_cognito_dart/src/model/auth_configuration.dart';
...
+      case ConfigurationEventType.configureFailed:
+        final castEvent = event as ConfigureFailed;
+        emit(ConfigurationState.failure(castEvent.exception));
+        await onConfigureFailed(castEvent);
+        return;
+    }
+  }
+
+  @override
+  ConfigurationState? resolveError(Object error, [StackTrace? st]) {
```
[the remark this answers is not in the document]
Comment (dnys1) [reply packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/configuration_state_machine.dart:67]: Good question - kind of an oversight, but it's also not straightforward what to do with it. Let me give it some thought and I'll put up a separate PR for this and include the new AWSResult as well.

File: packages/amplify_core/test/state_machine/state_machine_test.dart [hunk 3]
```diff
@@ -59,7 +59,7 @@ void main() {
       });
 
       test('can listen multiple times to other machines', () async {
```
Comment (Jordan-Nelson) [inline packages/amplify_core/test/state_machine/state_machine_test.dart:61]: Should we add a test for the queuing/completion behavior? Something like:

```dart
test('can queue events and process them in order', () async {
  final tryWork1 = stateMachine.accept(const MyEvent(MyType.tryWork));
  final delegate = stateMachine.accept(
    const MyEvent(MyType.delegateWork),
  );
  final tryWork2 = stateMachine.accept(const MyEvent(MyType.tryWork));
  await expectLater(
    stateMachine.stream,
    emitsInOrder([
      const MyState(MyType.tryWork),
      const MyState(MyType.error),
      const MyState(MyType.delegateWork),
      const WorkerState(WorkType.doWork),
      const WorkerState(WorkType.success),
      const MyState(MyType.success),
      const MyState(MyType.tryWork),
      const MyState(MyType.error),
    ]),
  );
  expect(await tryWork1.completed, const MyState(MyType.error));
  expect(await delegate.completed, const MyState(MyType.success));
  expect(await tryWork2.completed, const MyState(MyType.error));
});
```

**Review by Jordan-Nelson — APPROVED** [review]
Looks good!

**Pull request merged**